hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -927,81 +882,38 @@ public void onListenerRegister() {
 
     private void registerInvalidationListener() {
         if (nearCache != null && nearCache.isInvalidateOnChange()) {
-            ClientClusterService clusterService = clientContext.getClusterService();
-            nearCacheMembershipRegistrationId =
-                    clusterService.addMembershipListener(new NearCacheMembershipListener());
-            Collection<Member> memberList = clusterService.getMemberList();
-            for (Member member : memberList) {
-                addInvalidationListener(member);
-            }
-        }
-    }
-
-    private void removeInvalidationListener() {
-        if (nearCache != null && nearCache.isInvalidateOnChange()) {
-            String registrationId = nearCacheMembershipRegistrationId;
-            ClientClusterService clusterService = clientContext.getClusterService();
-            if (registrationId != null) {
-                clusterService.removeMembershipListener(registrationId);
-            }
-            Collection<Member> memberList = clusterService.getMemberList();
-            for (Member member : memberList) {
-                removeInvalidationListener(member, true);
-            }
-        }
-    }
-
-    private void addInvalidationListener(Member member) {
-        if (nearCacheInvalidationListeners.containsKey(member)) {
-            return;
-        }
-        try {
-            ClientMessage request = CacheAddInvalidationListenerCodec.encodeRequest(nameWithPrefix);
             Client client = clientContext.getClusterService().getLocalClient();
             EventHandler handler = new NearCacheInvalidationHandler(client);
-            HazelcastClientInstanceImpl clientInstance = (HazelcastClientInstanceImpl) clientContext.getHazelcastInstance();
-            Address address = member.getAddress();
-            ClientInvocation invocation = new ClientListenerInvocation(clientInstance, handler, request, address,
-                    new ClientMessageDecoder() {
-                        @Override
-                        public <T> T decodeClientMessage(ClientMessage clientMessage) {
-                            return (T) CacheAddInvalidationListenerCodec.decodeResponse(clientMessage).response;
-                        }
-                    });
-            Future<ClientMessage> future = invocation.invoke();
-            String registrationId = CacheAddInvalidationListenerCodec.decodeResponse(future.get()).response;
-            clientContext.getListenerService().registerListener(registrationId, request.getCorrelationId());
+            nearCacheMembershipRegistrationId = clientContext.getListenerService().startListening(new ListenerMessageCodec() {
```

Comment:
can be improved readability by creating codec first then calling `startListening`
you can even take creation to a method